File: Python3/2861_maximum_number_of_alloys.py
#  https://leetcode.cn/problems/maximum-number-of-alloys/

class Solution:
    def maxNumberOfAlloys(self, n, k, budget, composition, stock, cost):
        """
        Inputs:
            n, k, budget: int
            composition: list[list[int]]
            stock, cost: list[int]
        Outputs:
            res: int
        """
        res = 0
        left, right = 0, min(stock) + budget
        while left <= right:
            mid = (left + right) // 2
            valid = False
            for i in range(k):
                spend = 0
                for j in range(n):
                    spend += max(composition[i][j]*mid-stock[j], 0) * cost[j]
                if spend <= budget:
                    valid = True
                    break
            if valid:
                left = mid + 1
                res = mid
            else:
                right = mid - 1
        return res

File: Cpp/2861_maximum_number_of_alloys.cpp
// https://leetcode.cn/problems/maximum-number-of-alloys/
        
class Solution {
public:
    int maxNumberOfAlloys(int n, int k, int budget, vector<vector<int>>& composition, vector<int>& stock, vector<int>& cost) {
        int res = 0, left = 0, right = stock[0];
        for (int s: stock) {
            right = min(right, s);
        }
        right += budget;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            bool valid = false;
            for (auto &com: composition) {
                long long spend = 0;
                for (int i = 0; i < n; ++i) {
                    spend += max((long long)com[i]*mid-stock[i], 0LL) * cost[i];
                }
                if (spend <= budget) {
                    valid = true;
                    break;
                }
            }
            if (valid) {
                left = mid + 1;
                res = mid;
            }
            else {
                right = mid - 1;
            }
        }
        return res;
    }
};
